File: javascript/socketd/src/socketd/transport/core/Processor.ts
import {Listener} from "./Listener";
import type {ChannelInternal} from "./Channel";
import {Frame} from "./Frame";
import {StreamInternal} from "../stream/Stream";
import {ChannelAssistant} from "./ChannelAssistant";

/**
 * 处理器
 *
 * @author noear
 * @since 2.0
 */
export interface Processor {
    /**
     * 设置监听器
     */
    setListener(listener: Listener);

    /**
     * 发送帧
     * */
    sendFrame<S>(channel: ChannelInternal, frame: Frame, channelAssistant:ChannelAssistant<S>, target:S );

    /**
     * 接收帧
     */
    reveFrame(channel: ChannelInternal, frame: Frame);

    /**
     * 打开时
     *
     * @param channel 通道
     */
    onOpen(channel: ChannelInternal);

    /**
     * 收到消息时
     *
     * @param channel 通道
     * @param frame   帧
     */
    onMessage(channel: ChannelInternal, frame: Frame);

    /**
     * 收到签复时
     *
     * @param channel 通道
     * @param frame   帧
     * @param stream  流
     */
    onReply(channel: ChannelInternal, frame: Frame, stream:StreamInternal<any> | null);


    /**
     * 关闭时
     *
     * @param channel 通道
     */
    onClose(channel: ChannelInternal);


    /**
     * 出错时
     *
     * @param channel 通道
     * @param error   错误信息
     */
    onError(channel: ChannelInternal, error: Error);


    /**
     * 执行关闭通知
     *
     * @param channel 通道
     */
    doCloseNotice(channel: ChannelInternal);
}

File: python/socketd/socketd/transport/core/Processor.py
from abc import ABC, abstractmethod
from typing import TypeVar, Callable

from socketd.transport.core import Listener
from socketd.transport.core.Channel import Channel
from socketd.transport.core.ChannelAssistant import ChannelAssistant
from socketd.transport.core.ChannelInternal import ChannelInternal
from socketd.transport.core.Frame import Frame
from socketd.transport.stream.Stream import StreamInternal

S = TypeVar("S")

class Processor(ABC):

    @abstractmethod
    def set_listener(self, listener: Listener) -> None:
        pass

    @abstractmethod
    def send_frame(self, channel: ChannelInternal, frame: Frame, channelAssistant: ChannelAssistant[S], target: S):
        pass

    @abstractmethod
    def reve_frame(self, channel: Channel, frame: Frame):
        pass

    @abstractmethod
    def on_open(self, channel: ChannelInternal):
        pass

    @abstractmethod
    def on_message(self, channel: ChannelInternal, frame: Frame):
        pass

    @abstractmethod
    def on_reply(self, channel: ChannelInternal, frame: Frame, stream: StreamInternal) -> None:
        ...

    @abstractmethod
    def on_close(self, channel: ChannelInternal):
        pass

    @abstractmethod
    def on_error(self, channel: ChannelInternal, error):
        pass
    @abstractmethod
    def do_close_notice(self, channel: ChannelInternal):
        pass
